packages/date-time/src/components/Calendar/CalendarYear/CalendarYear.base.tsx

```diff
@@ -0,0 +1,458 @@
+import { ICalendarIconStrings } from '../Calendar.types';
+
+const getClassNames = classNamesFunction<ICalendarYearStyleProps, ICalendarYearStyles>();
+
+const CELL_COUNT = 12;
+
+const DefaultCalendarYearStrings: ICalendarYearStrings = {
+  prevRangeAriaLabel: undefined,
+  nextRangeAriaLabel: undefined
+};
+
+const DefaultNavigationIcons: ICalendarIconStrings = {
+  leftNavigation: 'Up',
+  rightNavigation: 'Down',
+  closeIcon: 'CalculatorMultiply'
+};
+
+interface ICalendarYearGridCell {
+  focus(): void;
+}
+
+class CalendarYearGridCell extends React.Component<ICalendarYearGridCellProps, {}> implements ICalendarYearGridCell {
+  private _buttonRef: HTMLButtonElement;
+
+  public focus(): void {
+    if (this._buttonRef) {
+      this._buttonRef.focus();
+    }
+  }
+
+  public render(): JSX.Element {
+    const { styles, theme, className, highlightCurrentYear, highlightSelectedYear, year, selected, disabled, onSelectYear } = this.props;
+
+    const classNames = getClassNames(styles, {
+      theme: theme!,
+      className: className,
+      highlightCurrentYear: highlightCurrentYear,
+      highlightSelectedYear: highlightSelectedYear
+    });
+
+    return (
+      <button
+        className={css('ms-DatePicker-yearOption', classNames.yearButton, {
+          [`ms-DatePicker-day--highlighted ${classNames.selectedYear}`]: selected,
+          [`ms-DatePicker-yearOption--disabled ${classNames.disabledStyle}`]: disabled
+        })}
+        type="button"
+        role="gridcell"
+        onClick={!disabled && onSelectYear ? this._onClick : undefined}
+        onKeyDown={!disabled && onSelectYear ? this._onKeyDown : undefined}
+        disabled={disabled}
+        aria-label={String(year)}
+        aria-selected={selected}
+        ref={this._onButtonRef}
+      >
+        {this._onRenderYear()}
+      </button>
+    );
+  }
+
+  private _onRenderYear = () => {
+    const { year, onRenderYear } = this.props;
+    if (onRenderYear) {
+      return onRenderYear(year);
+    }
+    return year;
+  };
+
+  private _onButtonRef = (ref: HTMLButtonElement) => {
+    this._buttonRef = ref;
+  };
+
+  private _onClick = () => {
+    if (this.props.onSelectYear) {
+      this.props.onSelectYear(this.props.year);
+    }
+  };
+
+  private _onKeyDown = (ev: React.KeyboardEvent<HTMLElement>) => {
+    if (this.props.onSelectYear && ev.which === KeyCodes.enter) {
+      this.props.onSelectYear(this.props.year);
+    }
+  };
+}
+
+interface ICalendarYearGrid {
+  focus(): void;
+}
+class CalendarYearGrid extends React.Component<ICalendarYearGridProps, {}> implements ICalendarYearGrid {
+  private _selectedCellRef: CalendarYearGridCell;
+  private _currentCellRef: CalendarYearGridCell;
+
+  public focus(): void {
+    if (this._selectedCellRef) {
+      this._selectedCellRef.focus();
+    } else if (this._currentCellRef) {
+      this._currentCellRef.focus();
+    }
+  }
+  public render(): JSX.Element {
```

Comment:
>public render(): JSX.Element { [](start = 2, length = 30)

I notice that we have eight render functions for the eight different react components in this one file.  

I feel like some of these could simply be turned into functions or moved out, or organized better.

Reply:
Since this PR is porting an existing feature into this package, I didn't do much up-front refactoring. We have to decide (and we have been discussing offline) what's the best approach to bring the three pickers in this component (day, month, year) in harmony w.r.t. composition.

Pretty sure there will be another pass for that work.

Reply:
Since this an initial PR to get the feature ported from the existing control, I didn't do the bigger refactor required to bring the various pickers (day, month, now year) in sync w.r.t. composition. We have been discussing various approaches to this offline, and pretty sure that work will follow soon.